Scanner: Frequent constraint violations on insertion to ProvenanceBasedPostgresStorage

When running scans against random projects with latest ORT, I usually see multiple occurences
of constraint violations in the logs. I'm running ORT's `scanner` command only on one machine, and no other `scanner` command does access the database in parallel.

```
WARN  Exposed - Transaction attempt #0 failed: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "provenance_scan_results_vcs_type_vcs_url_vcs_revision_scanner_n"
  Detail: Key (vcs_type, vcs_url, vcs_revision, scanner_name, scanner_version, scanner_configuration)=(Git, https://github.com/janl/mustache.js.git, 813e273a658677852ab37e6f47c98a9d9352ccde, ScanCode, 31.2.1, --copyright --license --info --strip-root --timeout 300 --json-pp) already exists.. Statement(s): INSERT INTO provenance_scan_results (artifact_hash, artifact_url, scan_summary, scanner_configuration, scanner_name, scanner_version, vcs_revision, vcs_type, vcs_url) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
org.jetbrains.exposed.exceptions.ExposedSQLException: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "provenance_scan_results_vcs_type_vcs_url_vcs_revision_scanner_n"
  Detail: Key (vcs_type, vcs_url, vcs_revision, scanner_name, scanner_version, scanner_configuration)=(Git, https://github.com/janl/mustache.js.git, 813e273a658677852ab37e6f47c98a9d9352ccde, ScanCode, 31.2.1, --copyright --license --info --strip-root --timeout 300 --json-pp) already exists.
	at org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:64)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:141)
	at org.jetbrains.exposed.sql.Transaction.exec(Transaction.kt:127)
	at org.jetbrains.exposed.sql.statements.Statement.execute(Statement.kt:28)
	at org.jetbrains.exposed.sql.QueriesKt.insert(Queries.kt:71)
	at org.ossreviewtoolkit.scanner.storages.ProvenanceBasedPostgresStorage$write$1.invoke(ProvenanceBasedPostgresStorage.kt:144)
	at org.ossreviewtoolkit.scanner.storages.ProvenanceBasedPostgresStorage$write$1.invoke(ProvenanceBasedPostgresStorage.kt:143)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.inTopLevelTransaction$run(ThreadLocalTransactionManager.kt:215)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.access$inTopLevelTransaction$run(ThreadLocalTransactionManager.kt:1)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt$inTopLevelTransaction$1.invoke(ThreadLocalTransactionManager.kt:241)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.keepAndRestoreTransactionRefAfterRun(ThreadLocalTransactionManager.kt:249)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.inTopLevelTransaction(ThreadLocalTransactionManager.kt:240)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt$transaction$1.invoke(ThreadLocalTransactionManager.kt:190)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.keepAndRestoreTransactionRefAfterRun(ThreadLocalTransactionManager.kt:249)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction(ThreadLocalTransactionManager.kt:160)
	at org.jetbrains.exposed.sql.transactions.ThreadLocalTransactionManagerKt.transaction(ThreadLocalTransactionManager.kt:146)
	at org.ossreviewtoolkit.model.utils.DatabaseUtils.transaction(DatabaseUtils.kt:120)
	at org.ossreviewtoolkit.scanner.storages.ProvenanceBasedPostgresStorage.write(ProvenanceBasedPostgresStorage.kt:143)
	at org.ossreviewtoolkit.scanner.Scanner.storeProvenanceScanResult(Scanner.kt:549)
	at org.ossreviewtoolkit.scanner.Scanner.runPathScanners(Scanner.kt:383)
	at org.ossreviewtoolkit.scanner.Scanner.scan(Scanner.kt:155)
	at org.ossreviewtoolkit.scanner.Scanner$scan$4.invokeSuspend(Scanner.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:284)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.runScanners(ScannerCommand.kt:232)
	at org.ossreviewtoolkit.cli.commands.ScannerCommand.run(ScannerCommand.kt:159)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:198)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:211)
	at com.github.ajalt.clikt.parsers.Parser.parse(Parser.kt:18)
	at com.github.ajalt.clikt.core.CliktCommand.parse(CliktCommand.kt:400)
	at com.github.ajalt.clikt.core.CliktCommand.parse$default(CliktCommand.kt:397)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:415)
	at com.github.ajalt.clikt.core.CliktCommand.main(CliktCommand.kt:440)
	at org.ossreviewtoolkit.cli.OrtMainKt.main(OrtMain.kt:76)
Caused by: org.postgresql.util.PSQLException: ERROR: duplicate key value violates unique constraint "provenance_scan_results_vcs_type_vcs_url_vcs_revision_scanner_n"
  Detail: Key (vcs_type, vcs_url, vcs_revision, scanner_name, scanner_version, scanner_configuration)=(Git, https://github.com/janl/mustache.js.git, 813e273a658677852ab37e6f47c98a9d9352ccde, ScanCode, 31.2.1, --copyright --license --info --strip-root --timeout 300 --json-pp) already exists.
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2676)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2366)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:356)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:496)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:413)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:190)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:152)
	at com.zaxxer.hikari.pool.ProxyPreparedStatement.executeUpdate(ProxyPreparedStatement.java:61)
	at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeUpdate(HikariProxyPreparedStatement.java)
	at org.jetbrains.exposed.sql.statements.jdbc.JdbcPreparedStatementImpl.executeUpdate(JdbcPreparedStatementImpl.kt:26)
	at org.jetbrains.exposed.sql.statements.InsertStatement.execInsertFunction(InsertStatement.kt:123)
	at org.jetbrains.exposed.sql.statements.InsertStatement.executeInternal(InsertStatement.kt:131)
	at org.jetbrains.exposed.sql.statements.InsertStatement.executeInternal(InsertStatement.kt:12)
	at org.jetbrains.exposed.sql.statements.Statement.executeIn$exposed_core(Statement.kt:62)
	... 39 common frames omitted
```

--- a/scanner/src/main/kotlin/Scanner.kt
+++ b/scanner/src/main/kotlin/Scanner.kt
@@ -79,6 +79,15 @@ class Scanner(
         require(scannerWrappers.isNotEmpty() && scannerWrappers.any { it.value.isNotEmpty() }) {
             "At least one ScannerWrapper must be provided."
         }
+
+        scannerWrappers.values.flatten().distinct().forEach { scannerWrapper ->
+            scannerWrapper.criteria?.let { criteria ->
+                require(criteria.matches(scannerWrapper.details)) {
+                    "The scanner details of scanner '${scannerWrapper.details.name}' must satisfy the configured " +
+                        "criteria for looking up scan storage entries."
+                }
+            }
+        }
     }
 
     private val archiver = scannerConfig.archive.createFileArchiver()

--- a/scanner/src/main/kotlin/ScannerCriteria.kt
+++ b/scanner/src/main/kotlin/ScannerCriteria.kt
@@ -82,16 +82,19 @@ data class ScannerCriteria(
 
         /**
          * The name of the property defining the regular expression for the scanner name as part of [ScannerCriteria].
+         * The [scanner details][ScannerDetails] of the corresponding scanner must match the criteria.
          */
         const val PROP_CRITERIA_NAME = "regScannerName"
 
         /**
-         * The name of the property defining the minimum version of the scanner as part of [ScannerCriteria].
+         * The name of the property defining the minimum version of the scanner as part of [ScannerCriteria]. The
+         * [scanner details][ScannerDetails] of the corresponding scanner must match the criteria.
          */
         const val PROP_CRITERIA_MIN_VERSION = "minVersion"
 
         /**
-         * The name of the property defining the maximum version of the scanner as part of [ScannerCriteria].
+         * The name of the property defining the maximum version of the scanner as part of [ScannerCriteria]. The
+         * [scanner details][ScannerDetails] of the corresponding scanner must match the criteria.
          */
         const val PROP_CRITERIA_MAX_VERSION = "maxVersion"
 

